perceval/backends/mediawiki.py

```diff
@@ -0,0 +1,517 @@
+
+        mediawiki_version = self.client.get_version()
+        logger.info("MediaWiki version: %s", mediawiki_version)
+
+        if (mediawiki_version[0] == 1 and mediawiki_version[1] >= 27) or \
+            mediawiki_version[0] > 1:
+            fetcher = self.__fetch_1_27(from_date)
+        else:
+            fetcher = self.__fetch_pre1_27(from_date)
+
+        for page_reviews in fetcher:
+            yield page_reviews
+
+
+    @metadata
+    def fetch_from_cache(self):
+        """Fetch the pages from the cache.
+
+        :returns: a generator of pages
+
+        :raises CacheError: raised when an error occurs accessing the
+            cache
+        """
+        if not self.cache:
+            raise CacheError(cause="cache instance was not provided")
+
+        cache_items = self.cache.retrieve()
+
+        pages_dicts = ['allrevisions','allpages','recentchanges']
+
+        # Namespaces -> pages -> revisions
+        next(cache_items) # First namespaces, we don't need them
+        for items in cache_items:
+            data_json = json.loads(items)
+            for pages_dict in pages_dicts:
+                if pages_dict in data_json['query']:
+                    pages_json = data_json['query'][pages_dict]
+                    for page in pages_json:
+                        page_reviews = self.__build_page_reviews(page, json.loads(next(cache_items)))
+                        yield page_reviews
+
+    @staticmethod
+    def metadata_id(item):
+        """Extracts the identifier from a wiki page."""
+        return str(item['pageid'])
+
+    @staticmethod
+    def metadata_updated_on(item):
+        """Extracts the update field from a MediaWiki item.
+
+        The timestamp is extracted from 'timestamp' field.
+        This date is a UNIX timestamp but needs to be converted to
+        a float value.
+
+        :param item: item generated by the backend
+
+        :returns: a UNIX timestamp
+        """
+        updated = None
+        if item['update']:
+            updated = float(item['update'])
+        return updated
+
+
+    def __get_max_date(self, reviews):
+        """"Get the max date in unixtime format from reviews"""
+        max_ts = 0
+        for review in reviews:
+            ts = str_to_datetime(review['timestamp'])
+            ts = datetime_to_utc(ts)
+            if ts.timestamp() > max_ts:
+                max_ts = ts.timestamp()
+        return max_ts
+
+
+    def __fetch_1_27(self, from_date=None):
+        """Fetch the pages from the backend url for MediaWiki >=1.27
+
+        The method retrieves, from a MediaWiki url, the
+        wiki pages.
+
+        :returns: a generator of pages
+        """
+
+        logger.info("Looking for pages at url '%s'", self.url)
+
+        self._purge_cache_queue()
+        npages = 0  # number of pages processed
+
+        # The pages are organized in content namespaces
+        raw_namespaces = self.client.get_namespaces()
```

Comment:
This code is the same that the one you use on line 280. This should be moved to a method.